posthog/hogql/printer.py

```diff
@@ -1058,6 +1059,26 @@ def visit_property_type(self, type: ast.PropertyType):
                     property_sql = self._print_identifier(materialized_column)
                     property_sql = f"{self.visit(field_type.table_type)}.{property_sql}"
                     materialized_property_sql = property_sql
+                else:
+                    property_name = str(type.chain[0])
+                    # For now, we're assuming that properties are in either no
+                    # groups or one group, so just using the first group
+                    # returned is fine. If we start putting properties in
+                    # multiple groups, this should be revisited to find the
+                    # optimal set (i.e. smallest set) of groups to read from.
+                    for property_group_column in property_groups.get_property_group_columns(
+                        table_name, field_name, property_name
+                    ):
+                        printed_column = (
+                            f"{self.visit(field_type.table_type)}.{self._print_identifier(property_group_column)}"
+                        )
+                        printed_property_name = self.context.add_value(property_name)
+                        # If the key we're looking for doesn't exist in the map
+                        # for this property group, an empty string (the default
+                        # value for the `String` type) is returned. Since that
+                        # is a valid property value, we need to check it here.
+                        materialized_property_sql = f"has({printed_column}, {printed_property_name}) ? {printed_column}[{printed_property_name}] : null"
```

Comment:
Note to self for next week: this `has` expression causes the bloom filter indices to be skipped when a property ends up in the `WHERE` clause of a query:

```
EXPLAIN indexes = 1
SELECT count()
FROM events
WHERE (events.team_id = 1) AND ifNull(if(has(events.properties_group_custom, 'file_type'), events.properties_group_custom['file_type'], NULL) = 'audio/vorbis', 0)
LIMIT 0, 101

Query id: d4bf9ebb-5cb0-4f4e-a218-a163621af497

┌─explain────────────────────────────────────────────┐
│ Expression ((Projection + Before ORDER BY))        │
│   Limit (preliminary LIMIT (without OFFSET))       │
│     Aggregating                                    │
│       Expression (Before GROUP BY)                 │
│         ReadFromMergeTree (default.sharded_events) │
│         Indexes:                                   │
│           MinMax                                   │
│             Condition: true                        │
│             Parts: 10/10                           │
│             Granules: 13/13                        │
│           Partition                                │
│             Condition: true                        │
│             Parts: 10/10                           │
│             Granules: 13/13                        │
│           PrimaryKey                               │
│             Keys:                                  │
│               team_id                              │
│             Condition: (team_id in [1, 1])         │
│             Parts: 10/10                           │
│             Granules: 13/13                        │
└────────────────────────────────────────────────────┘

20 rows in set. Elapsed: 0.007 sec.
```

versus

```
EXPLAIN indexes = 1
SELECT count()
FROM events
WHERE (events.team_id = 1) AND ((events.properties_group_custom['file_type']) = 'audio/vorbis')
LIMIT 0, 101

Query id: 772d0b63-a13a-4f6f-878e-c42f553b608d

┌─explain─────────────────────────────────────────────┐
│ Expression ((Projection + Before ORDER BY))         │
│   Limit (preliminary LIMIT (without OFFSET))        │
│     Aggregating                                     │
│       Expression (Before GROUP BY)                  │
│         ReadFromMergeTree (default.sharded_events)  │
│         Indexes:                                    │
│           MinMax                                    │
│             Condition: true                         │
│             Parts: 10/10                            │
│             Granules: 13/13                         │
│           Partition                                 │
│             Condition: true                         │
│             Parts: 10/10                            │
│             Granules: 13/13                         │
│           PrimaryKey                                │
│             Keys:                                   │
│               team_id                               │
│             Condition: (team_id in [1, 1])          │
│             Parts: 10/10                            │
│             Granules: 13/13                         │
│           Skip                                      │
│             Name: properties_group_custom_keys_bf   │
│             Description: bloom_filter GRANULARITY 1 │
│             Parts: 9/10                             │
│             Granules: 12/13                         │
│           Skip                                      │
│             Name: properties_group_custom_values_bf │
│             Description: bloom_filter GRANULARITY 1 │
│             Parts: 2/9                              │
│             Granules: 2/12                          │
└─────────────────────────────────────────────────────┘

30 rows in set. Elapsed: 0.012 sec. 
```

Reply:
The draft for optimizing these expressions that impact property group index usage is here: #24381